src/db/tests/db.rs

```diff
@@ -632,14 +639,79 @@ async fn get_collection_usage() -> Result<()> {
 
     let sizes = db.get_collection_usage(hid(uid)).await?;
     assert_eq!(sizes, expected);
+    let sum = expected.values().sum::<i64>();
     let total = db.get_storage_usage(hid(uid)).await?;
-    assert_eq!(total, expected.values().sum::<i64>() as u64);
+    assert_eq!(total, sum as u64);
+    let settings = test_settings();
+    if settings.enable_quota {
+        let collection_id = db.get_collection_id("bookmarks".to_owned()).await?;
+        let quota = db
+            .get_quota_usage(params::GetQuotaUsage {
+                user_id: HawkIdentifier::new_legacy(uid as u64),
+                collection: "ignored".to_owned(),
+                collection_id,
+            })
+            .await?;
+        assert_eq!(
+            &(quota.total_bytes as i64),
+            expected.get("bookmarks").unwrap()
+        );
+        assert_eq!(quota.count, 5); // 3 collections, 5 records
+    }
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_quota() -> Result<()> {
+    let settings = crate::settings::test_settings();
+
+    if !settings.enable_quota {
+        debug!("Skipping test");
+        return Ok(());
+    }
+
+    let pool = db_pool(None).await?;
+    let mut db = test_db(pool.as_ref()).await?;
+
+    let uid = 5;
+    let coll = "bookmarks";
+
+    let size = 5000;
+    let payload = thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(size)
+        .collect::<String>();
+    db.set_quota(false, 0);
+
+    // These should work
+    db.put_bso(pbso(uid, coll, "100", Some(&payload), None, None))
+        .await?;
+    db.put_bso(pbso(uid, coll, "101", Some(&payload), None, None))
+        .await?;
+
+    db.set_quota(true, size * 2);
+
+    // Allow the put, but calculate the quota
+    db.put_bso(pbso(uid, coll, "102", Some(&payload), None, None))
+        .await?;
+    // this should fail, since the user is already over the quota
+    let result = db
+        .put_bso(pbso(uid, coll, "103", Some(&payload), None, None))
+        .await;
+    assert!(result.is_err());
     Ok(())
 }
 
 #[tokio::test]
 async fn get_collection_counts() -> Result<()> {
-    let pool = db_pool().await?;
+    let settings = crate::settings::test_settings();
+
+    if !settings.enable_quota {
```

Comment:
I don't think we want to skip this one, though, it's existed for a while now

Reply:
ah, doh. sorry. got carried away.